DOC Document changes to CLI interaction (#571)
silverstripe/developer-docs: en/02_Developer_Guides/17_CLI/02_Hybrid_Commands.md

> ---

Reviewer: Rename file to 02_Poly_Commands.md

Reviewer 2: Renamed to `02_PolyCommand.md` which is what all the links to this page assume the name is.

> [!TIP]
> See [the symfony/console input documentation](https://symfony.com/doc/current/console/input.html#using-command-options) for more specific details about using `InputInterface` and `InputOption`. Note that we use the [`getOptions()`](api:SilverStripe\PolyExecution\PolyCommand::getOptions()) method instead of `configure()` to return an array of `InputOption` objects. `PolyCommand` does not allow arguments.
>
> See [the symfony/console colouring documentation](https://symfony.com/doc/current/console/coloring.html) for information about styling output. Note that all output (including outputting for HTTP requests) as well as the command description and help info can use the `symfony/console` styling format.

```php
namespace App\Cli\Command;

use SilverStripe\PolyExecution\PolyCommand;
use SilverStripe\PolyExecution\PolyOutput;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;

class MyPolyCommand extends PolyCommand
{

>     protected static string $commandName = 'app:my-command';

Reviewer: ```suggestion
    protected static string $commandName = 'app:MyPolyCommand';
```
As mentioned [here](https://github.com/silverstripe/silverstripe-framework/issues/11341#issuecomment-2367027143) I think we should be encouraging the ClassName style naming rather than url style

    protected string $title = 'My command';

    protected static string $description = 'A command that does something';

    public function run(InputInterface $input, PolyOutput $output): int
    {
        if ($input->getOption('do-action')) {
            $output->writeln('Doing something...');
        }
        return Command::SUCCESS;
    }

    public function getOptions(): array
    {
        return [
            new InputOption('do-action', null, InputOption::VALUE_NONE, 'do something specific'),
        ];
    }
}
```

> [!NOTE]
> You can also optionally implement the static [`getHelp()`](api:SilverStripe\PolyExecution\PolyCommand::getHelp()) method to provide additional context about the command. This helps to keep the description short.

Once you've set up your command in PHP, you can add an HTTP route to is using the regular [`Director` routing rules](/developer_guides/controllers/routing/), and add it to Sake with the [`Sake.commands`](api:SilverStripe\Cli\Sake->commands) configuration property. Note that this isn't necessary for the special subclasses mentioned in the sections below.

```yml
---
Name: polycommands
---
SilverStripe\Control\Director:
  rules:

>     my-command: 'App\Cli\Command\MyPolyCommand'

Reviewer: Is this now necessary? Or does protected static string $commandName provide a segment for dev/tasks the way that private static $segment used to?

Reviewer 2: This is an optional way that custom poly commands can be routed for access via an HTTP request.
This is not necessary for `BuildTask` or `DevCommand`.

SilverStripe\Cli\Sake:
  commands:
    - 'App\Cli\Command\MyPolyCommand'
```

This command could then be used by visiting `https://www.example.com/my-command` in a browser or running `sake app:my-command` in CLI.

## Configuration

For every `PolyCommand` subclass, including the special kinds listed below, you can set the following configuration:

> - [`can_run_in_cli`](api:SilverStripe\PolyExecution\PolyCommand->can_run_in_cli): Whether the command can be run in CLI
> - [`can_run_in_browser`](api:SilverStripe\PolyExecution\PolyCommand->can_run_in_browser): Whether the command can be run in HTTP requests
> - [`permissions_for_browser_execution`](api:SilverStripe\PolyExecution\PolyCommand->permissions_for_browser_execution): An array of permissions a user must have to run the command in an HTTP request

Reviewer: ```suggestion
- [`can_run_in_cli`](api:SilverStripe\PolyExecution\PolyCommand->can_run_in_cli): Whether the command can be run in CLI
- [`can_run_in_browser`](api:SilverStripe\PolyExecution\PolyCommand->can_run_in_browser): Whether the command can be run in HTTP requests
- [`permissions_for_browser_execution`](api:SilverStripe\PolyExecution\PolyCommand->permissions_for_browser_execution): An array of permissions of which the user must have at least one have to run the command in an HTTP request
```
By default `Permission::check()` will use `any` not `all`

Reviewer 2: Clarified in a different way (based on the comment in the main issue I assumed there would be no suggestion here so I made the change before I saw this)

You can also set the `can_run_in_cli` and `can_run_in_browser` configuration properties as mentioned in [`PolyCommand`](#polycommand) above.

The API for a `BuildTask` is effectively identical to other `PolyCommand` subclasses. Note that instead of implementing `run()` directly, you should implement the `execute()` method. There is some logic in [`BuildTask::run()`](api:SilverStripe\Dev\BuildTask::run()) which will output the title of your command before calling `execute()`. Once your task has finished, it will also output whether it was successful or not (based on the return value) and how long it took.

> [!WARNING]
> Unlike other `PolyCommand` subclasses, you cannot include a namespace for tasks. This is because the command name is also used as the URL segment for HTTP execution of the task, and in the CLI the `tasks:` namespace is automatically applied.

```php
namespace App\Tasks;

use SilverStripe\Dev\BuildTask;
use SilverStripe\PolyExecution\PolyOutput;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;

class MyCustomTask extends BuildTask
{

>     protected static string $commandName = 'my-custom-task';

Reviewer: ```suggestion
    protected static string $commandName = 'MyCustomTask';
```
I've recommended we use a $commandName that matches the short classname for all BuildTasks, so we should probably recommend the same convention in our docs

Reviewer 2: Discussion about this is ongoing in the issue.

    protected string $title = 'My custom task';

    protected static string $description = 'A task that does something custom';

    protected function execute(InputInterface $input, PolyOutput $output): int
    {
        if ($input->getOption('do-action')) {
            $output->writeln('Doing something...');
        }
        return Command::SUCCESS;
    }

    public function getOptions(): array
    {
        return [
            new InputOption('do-action', null, InputOption::VALUE_NONE, 'do something specific'),
        ];
    }
}
```

You don't need to do anything to register a `BuildTask` - it will be given an HTTP route and be registered with Sake simply by existing.

The above command would be accessible via the browser at `/dev/tasks/my-custom-task` and via CLI with `sake tasks:my-custom-task`.

### Running regular tasks with cron

On a UNIX machine, you can typically run a scheduled task with a [cron job](https://en.wikipedia.org/wiki/Cron). Run
your task in as a cron job using `sake`.

For example, the following will run `MyTask` every minute.

```bash

> * * * * * /your/site/folder/vendor/bin/sake tasks:my-task

Reviewer: ```suggestion
For example, the following will run `MyCustomTask` every minute.

```bash
* * * * * /your/site/folder/vendor/bin/sake tasks:MyCustomTask
```

Reviewer 2: Discussion about this is ongoing in the issue.
